components/zeekopenbsm/src/openbsmconsumer.cpp

```diff
@@ -0,0 +1,363 @@
+      event.header.local_port = 0;
+    }
+    if (tok.tt.sockinet_ex32.family == 2) {
+      event.header.family = 2;
+    } else if (tok.tt.sockinet_ex32.family == 26) {
+      event.header.family = 10;
+    } else {
+      event.header.family = 0;
+    }
+
+    break;
+  }
+  case AUT_SOCKINET128: {
+    if (event.type == Event::Type::Bind) {
+      event.header.remote_address = "0.0.0.0";
+      event.header.remote_port = 0;
+      event.header.local_address = getIpFromToken(tok);
+      event.header.local_port =
+          static_cast<std::int64_t>(ntohs(tok.tt.sockinet_ex32.port));
+    } else {
+      event.header.remote_address = getIpFromToken(tok);
+      event.header.remote_port =
+          static_cast<std::int64_t>(ntohs(tok.tt.sockinet_ex32.port));
+      event.header.local_address = "0.0.0.0";
+      event.header.local_port = 0;
+    }
+    if (tok.tt.sockinet_ex32.family == 2) {
+      event.header.family = 2;
+    } else if (tok.tt.sockinet_ex32.family == 26) {
+      event.header.family = 10;
+    } else {
+      event.header.family = 0;
+    }
+    break;
+  }
+  }
+}
+
+Status
+OpenbsmConsumer::populateSocketEvent(Event &event,
+                                     const std::vector<tokenstr_t> &tokens) {
+
+  for (const auto &tok : tokens) {
+    if (tok.id == AUT_SOCKUNIX) {
+      // we filter unix sockets out
+      continue;
+    }
+    switch (tok.id) {
+    case AUT_HEADER32:
+    case AUT_HEADER64:
+    case AUT_HEADER32_EX:
+    case AUT_HEADER64_EX:
+      extractHeader(event, tok);
+      break;
+    case AUT_SUBJECT32:
+    case AUT_SUBJECT64:
+    case AUT_SUBJECT32_EX:
+    case AUT_SUBJECT64_EX:
+      extractSubject(event, tok);
+      break;
+    case AUT_RETURN32:
+    case AUT_RETURN64:
+      extractReturn(event, tok);
+      break;
+    case AUT_SOCKINET32:
+    case AUT_SOCKINET128:
+      extractSocketInet(event, tok);
+      break;
+    }
+  } // for loop
+
+  return Status::success();
+}
+
+Status OpenbsmConsumer::parseMessage() {
+
+  Status status;
+  Event event;
+
+  u_char *buffer = nullptr;
+
+  int reclen = au_read_rec(audit_pipe, &buffer);
+
+  if (reclen <= 0) {
+    return Status::failure("Could not openbsm fetch message");
+  }
+
+  tokenstr_t tok;
+  std::vector<tokenstr_t> tokens{};
+  tokens.reserve(12);
+
+  auto event_id = 0;
+  auto bytesread = 0;
+  while (bytesread < reclen) {
+    if (au_fetch_tok(&tok, buffer + bytesread, reclen - bytesread) == -1) {
+      break;
+    }
+    switch (tok.id) {
+    case AUT_HEADER32:
+      event_id = tok.tt.hdr32.e_type;
+      break;
+    case AUT_HEADER32_EX:
+      event_id = tok.tt.hdr32_ex.e_type;
+      break;
+    case AUT_HEADER64:
+      event_id = tok.tt.hdr64.e_type;
+      break;
+    case AUT_HEADER64_EX:
+      event_id = tok.tt.hdr64_ex.e_type;
+      break;
+    }
+    tokens.push_back(tok);
+    bytesread += tok.len;
+  }
+  if (event_filter_list.find(event_id) == event_filter_list.end()) {
```

Comment:
Doesn't this assume that the tokens we receive are ordered in a way where "interesting" tokens always preceed "uninteresting" ones? We would skip the rest of the function, even if we found some interesting ones.

Maybe I am misunderstanding what this does, but we also seem to reuse the `event_id` of the previous iteration if no case matches. Shouldn't we reset its value in a default case?

Reply:
So it works like this: each event (e.g. socket connect) generates multiple token strings. Each generated token string has a type that is defined by `type.id`. HEADER is a must token type that appears only once per event. In this case, `event_id` will only be assigned once in the loop since there is only one token string with HEADER type.
I hope this anwsers your first question as well. Let me know if it is still unclear.

Reply:
Thanks, that helps. Can we work on making the implementation more closely model what you explained? We could e.g., log a (fatal?) error if `au_fetch_tok` fails, move the storing of `tok` in `tokens` right after the fetch since they belong together, make `event_id` a `optional<u_int16_t>` and `assert(event_id)` before using it, and expand the comment for the early return with something capturing the guaranteed semantics.

Reply:
I added your suggestion in the lastest commit.